src/RequestCompression/RequestCompressionMiddleware.php

```diff
@@ -0,0 +1,168 @@
+            $command,
+            $operation,
+            $requestBodySize
+        )) {
+            return $nextHandler($command, $request);
+        }
+
+        $this->encodings = $compressionTrait['encodings'];
+        $request = $this->compressRequestBody($request);
+
+        return $nextHandler($command, $request);
+    }
+
+    private function compressRequestBody(
+        RequestInterface $request
+    ) {
+        $fn = $this->determineEncoding();
+        if (is_null($fn)) {
+            return $request;
+        }
+
+        $body = $request->getBody()->getContents();
+        $compressedBody = $fn($body);
+
+        return $request->withBody(Psr7\Utils::streamFor($compressedBody))
+            ->withHeader('content-encoding', $this->encoding);
+    }
+
+    private function determineEncoding()
+    {
+        foreach ($this->encodings as $encoding) {
+            if (isset($this->encodingMap[$encoding])) {
+                $this->encoding = $encoding;
+                return $this->encodingMap[$encoding];
+            }
+        }
+        return null;
+    }
+
+    private function shouldCompressRequestBody(
+        $compressionTrait,
+        $command,
+        $operation,
+        $requestBodySize
+    ){
+        if ($compressionTrait) {
+            if (isset($command['@disable_request_compression'])
+                && $command['@disable_request_compression'] === true
+            ) {
+                return false;
+            } elseif ($this->hasStreamingTraitWithoutRequiresLength($command, $operation)
+                || $requestBodySize >= $this->minimumCompressionSize
+            ) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private function hasStreamingTraitWithoutRequiresLength($command, $operation)
+    {
+        $hasStreaming = false;
+        $requiresLength = false;
+        foreach ($operation->getInput()->getMembers() as $name => $member) {
+            if (!empty($member['streaming']) && isset($command[$name])) {
+                $hasStreaming = true;
+                if (!empty($member['requiresLength'])) {
+                    $requiresLength = true;
+                }
+            }
+        }
+
+        return $hasStreaming && !$requiresLength;
+    }
+
+    private function determineMinimumCompressionSize($config) {
+        if (isset($config['request_min_compression_size_bytes'])) {
+            if (is_callable($config['request_min_compression_size_bytes'])) {
+                $minCompressionSz = $config['request_min_compression_size_bytes']()
+                    ->wait()
+                    ->getMinCompressionSize();
+            } else {
+                $minCompressionSz = $config['request_min_compression_size_bytes'];
+            }
+
+            if ($this->isValidCompressionSize($minCompressionSz)) {
+                return $minCompressionSz;
+            }
+        }
+
+        return self::REQUEST_MIN_COMPRESSION_SIZE_BYTES;
```

Comment:
When is this called?  If the user does specify a minimum size, then it will be converted into the config object (if it isn't already that type).  If a user does not, then this will be set to a config option by the default provider.  
What codepath will lead the if statement above to be false